engine/gamesys/src/gamesys/scripts/script_resource.cpp

```diff
@@ -1003,6 +1016,131 @@ static int SetTexture(lua_State* L)
+ *
+ * `mipmaps`
+ * : [type:integer] number of mipmaps of the texture
+ *
+ * `type`
+ * : [type:number] The texture type. Supported values:
+ *
+ * - `resource.TEXTURE_TYPE_2D`
+ * - `resource.TEXTURE_TYPE_CUBE_MAP`
+ * - `resource.TEXTURE_TYPE_2D_ARRAY`
+ *
+ * @examples
+ * Create a new texture and get the metadata from it
+ *
+ * ```lua
+ * function init(self)
+ *     -- create an empty texture
+ *     local tparams = {
+ *         width          = 128,
+ *         height         = 128,
+ *         type           = resource.TEXTURE_TYPE_2D,
+ *         format         = resource.TEXTURE_FORMAT_RGBA,
+ *     }
+ *
+ *     local my_texture_path = resource.create_texture("/my_texture.texturec", tparams)
+ *     local my_texture_info = resource.get_texture_info(my_texture_path)
+ *
+ *     -- my_texture_info now contains
+ *     -- {
+ *     --      handle = <the-numeric-handle>,
+ *     --      width = 128,
+ *     --      height = 128,
+ *     --      depth = 1
+ *     --      mipmaps = 1,
+ *     --      type = resource.TEXTURE_TYPE_2D
+ *     -- }
+ * end
+ * ```
+ *
+ * @examples
+ * Get the meta data from an atlas resource
+ *
+ * ```lua
+ * function init(self)
+ *     local my_atlas_info   = resource.get_atlas("/my_atlas.a.texturesetc")
+ *     local my_texture_info = resource.get_texture_info(my_atlas_info.texture)
+ *
+ *     -- my_texture_info now contains the information about the texture that is backing the atlas
+ * end
+ * ```
+ */
+static int GetTextureInfo(lua_State* L)
+{
+    int top = lua_gettop(L);
+    dmGraphics::HTexture texture_handle = 0;
+
+    if (lua_isnumber(L, 1))
+    {
+        texture_handle = lua_tonumber(L, 1);
+        if (!dmGraphics::IsAssetHandleValid(g_ResourceModule.m_GraphicsContext, texture_handle))
+        {
+            return luaL_error(L, "Texture handle is not valid.");
+        }
+    }
+    else
+    {
+        dmhash_t path_hash           = dmScript::CheckHashOrString(L, 1);
+        TextureResource* texture_res = (TextureResource*) CheckResource(L, g_ResourceModule.m_Factory, path_hash, "texturec");
+        texture_handle               = texture_res->m_Texture;
+
+        if (!dmGraphics::IsAssetHandleValid(g_ResourceModule.m_GraphicsContext, texture_handle))
+        {
+            return luaL_error(L, "Texture '%s' is not a valid texture handle.", dmHashReverseSafe64(path_hash));
+        }
+    }
+
+    uint32_t texture_width               = dmGraphics::GetTextureWidth(texture_handle);
+    uint32_t texture_height              = dmGraphics::GetTextureHeight(texture_handle);
+    uint32_t texture_depth               = dmGraphics::GetTextureDepth(texture_handle);
+    uint32_t texture_mipmaps             = dmGraphics::GetTextureMipmapCount(texture_handle);
+    dmGraphics::TextureType texture_type = dmGraphics::GetTextureType(texture_handle);
+
+    lua_newtable(L);
+
+    lua_pushnumber(L, texture_handle);
+    lua_setfield(L, -2, "handle");
```

Comment:
Is this really safe for all possible 64 bit signed integers?
Some of those bit patterns will represent NaN's etc, are we guaranteed to get the same number back? And does it work the same way for both Lua 5.1 and LuaJIT?

I'm wondering if using it as a light user data instead removes those questions altogether:
https://www.lua.org/manual/5.1/manual.html#lua_pushlightuserdata

Reply:
Hmm, that's a good point. The reason I opted for the 64b handle was to store the type information in the handle itself so we know what it's supposed to represent. I.e if you create a texture via the resource namespace and pass it to another system (render script most likely) and call render.enable_texture with the handle we need to be able to distinguish between a render target and a texture. Alternatively we can represent the handle as a 32b value and store the type information in the graphics adapters, but it's a less 'neat' solution since we need another indirection just to store that extra bit (or maybe bits in the future). Maybe I'm wrong but with lightuserdata I'm not sure we solve this issue anyway since we don't serialize that type in the message system right? And they are 32b on win32 so I guess the only thing it helps with in that case is to preserve the bits in case lua does something strange with integers